Added a new module that can manage rules in pg_hba files.

sebasmannem opened the pull request:
##### SUMMARY
Postgres pg_hba files are difficult to manage. They can best be related to iptables config:
- first lines hit, will be applied
- there are complexe datatypes involved, like ip addresses and subnets, which could have different notations for the same value.
For these reasons (order is important, complex datatypes), lineinfile doesn't really work well.

On the other hand, templating does work, but overwrite the whole file. Any changes which are there, but are not part of the template, are basically reverted. With Ansible, you can manage only parts of config, but not for pg_hba rules.

This module fixes that. You an state that a hba rule should be there and the modules makes sure it is there and even on the on a logical place, so that it hits, without overriding other rules (other than what should be logical).

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
pg_hba

##### ANSIBLE VERSION
```
ansible 2.4.1.0
  config file = None
  configured module search path = [u'/Users/smannem/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/Cellar/ansible/2.4.1.0/libexec/lib/python2.7/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.14 (default, Sep 25 2017, 09:53:22) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.37)]
```


##### ADDITIONAL INFORMATION
This module already was created in juli 2015, and used internally on many occasions. I wanted to commit it to the community,since then, but never took the time too jump through all hoops. It has always been on my personal github account, for anyone to use. I just hope to help others to manag postgres databases in a more transparant way.

After this change, you can
```
ansible -m pg_hba -a 'dest=/tmp/pg_hba.conf contype=host source=192.168.8.0 netmask=255.255.255.0 method=md5 backup=true order=sud state=present create=true'
```

Commit:
Added a new module that can manage rules in pg_hba files.

nerzhul commented:
@jctanner seems bot sent the same notification 3 times

sebasmannem commented:
@b6d @dschep @jensdepuydt @matburt @nerzhul : Can you tell me what the state is, that this merge request is in? Is it going to be merged? In what version? 2.5?

sebasmannem commented:
I'd like to be more active. I have some more ideaus:
- postgres_query, to run a query on a postgres database (same as shell, but then for all pg clusters)
- oracle_query, idem, but for oracle
- oratab, to manage (read and modify) oratab files
Let me know if there is any interest. For the oracle stuff, I need to create a test environment, but I'm happy to...

pilou- commented:
@sebasmannem about postgres_query there is an existing related pull-request: #25411.

pilou- commented on lib/ansible/modules/database/postgresql/pg_hba.py:
```diff
@@ -0,0 +1,769 @@
+
+# This file is part of Ansible
+#
+# Ansible is free software: you can redistribute it and/or modify
```
Use shorter licence declaration ([example](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/monitoring/grafana_plugin.py#L4)).

sebasmannem replied:
Fixed in new commit (coming up).

pilou- commented on lib/ansible/modules/database/postgresql/pg_hba.py:
```diff
@@ -0,0 +1,769 @@
+    description:
+      - Path to C(pg_hba) file to modify.
+    required: true
+    default: ''
```
required parameter don't have default value.

sebasmannem replied:
Fixed in new commit (coming up).

pilou- commented on lib/ansible/modules/database/postgresql/pg_hba.py:
```diff
@@ -0,0 +1,769 @@
+    default: false
+  owner:
+    description:
+      - Owner of the C(pg_hba) file (almost always postgres, or sometimes enterprisedb).
```
Formatting function `C()` could be used with `postgres` and `enterprisedb`.

sebasmannem replied:
Fixed in new commit (coming up).

sebasmannem replied:
And extends_documentation_fragment has even removed owner from the docstring entirely.

pilou- commented on lib/ansible/modules/database/postgresql/pg_hba.py:
```diff
@@ -0,0 +1,769 @@
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
```
`dest`, `owner`, `group`, `mode` parameters are present: common file arguments must be used instead:

-  use [`add_file_common_args=True`](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/crypto/openssl_csr.py#L472) parameter
- use [`extends_documentation_fragment`](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/crypto/openssl_csr.py#L151) ([doc](http://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html#documentation-fragments)).

sebasmannem replied:
Fixed in upcoming commit

sebasmannem replied:
Also changing source to src parameter

pilou- commented on lib/ansible/modules/database/postgresql/pg_hba.py:
```diff
@@ -0,0 +1,769 @@
+    description:
+      - Type of the rule. Use emptystring C('') if you don't want to change file, but only want to read contents.
+    required: true
+    default: "host"
```
`required` parameters don't have a default. Besides `required` isn't used in `argument_spec` for this parameter. I suggest to not require this parameter without any default value (then`None` would be the default value),meaning the default behavior would be to read contents unless `contype` is set.

sebasmannem replied:
Good notion. I have changed as you suggestedin new commit (coming up).

pilou- commented on lib/ansible/modules/database/postgresql/pg_hba.py:
```diff
@@ -0,0 +1,769 @@
+                ret['backup_file'] = pg_hba.last_backup
+
+    ret['pg_hba'] = pg_hba.get_rules()
+    ret['rc'] = 0
```
`rc` is [deprecated](https://github.com/ansible/ansible/blob/devel/lib/ansible/executor/task_executor.py#L572), don't use it.

pilou- commented on lib/ansible/modules/database/postgresql/pg_hba.py:
```diff
@@ -0,0 +1,769 @@
+        raise TouchError("Could not convert '{0}' form octal to int. Please specify a valid mode in octal form (e.a. 777, 640, etc.).".format(mode))
+    if mode > 511 or mode < 0:
+        raise TouchError("Please specify octal mode ({0}) between 000 and 777.".format(mode))
+    try:
```
`AnsibleModule.set_fs_attributes_if_different` should be used instead of the `touch` method (related to the comment below about _common file arguments_).

pilou- commented on lib/ansible/modules/database/postgresql/pg_hba.py:
```diff
@@ -0,0 +1,769 @@
+
+
+# See http://stackoverflow.com/questions/53497/regular-expression-that-matches-valid-ipv6-addresses for more info...
+IPV4SEG = '(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])'
```
[ipaddress](https://docs.python.org/3/library/ipaddress.html) Python module should be used instead in order to check/handle IP addresses. This module is included in Python 3.3 and backport exists for previous versions.

sebasmannem replied:
Yes, I like that module and I see the benefit in using that over my implementation. But I need some time on that.

sebasmannem replied:
Modified in latest commit.

pilou- commented on lib/ansible/modules/database/postgresql/pg_hba.py:
```diff
@@ -0,0 +1,769 @@
+            self.create = False
+        self.unchanged()
+        # self.databases will be update by add_rule and gives some idea of the number of databases (at least that are handled by this pg_hba)
+        self.databases = set(['postgres', 'template0', 'template1'])
```
`postgres` database could have been deleted.

pilou- commented on lib/ansible/modules/database/postgresql/pg_hba.py:
```diff
@@ -0,0 +1,769 @@
+        return tuple(ret)
+
+    def is_empty(self):
+        if len(self.rules) == 0:
```
The if/else block be replaced by `return not self.rules`.

sebasmannem replied:
I have enabled pylint and he tells me the same. Will fix all errors. This is fixed.

pilou- commented on lib/ansible/modules/database/postgresql/pg_hba.py:
```diff
@@ -0,0 +1,769 @@
+        return result
+
+    def changed(self):
+        if self.diff['before'] or self.diff['after']:
```
this could be `return self.diff['before'] or self.diff['after']`

sebasmannem replied:
fixed

pilou- reviewed (commented):
Tests should be added. There are already some integration tests related to PostgreSQL in test/integration/targets/postgresql directory.

pilou- commented on test/integration/targets/postgresql/tasks/pg_hba.yml:
```diff
@@ -0,0 +1,35 @@
+  file:
+    dest=/tmp/pg_hba.conf
+    state=absent
+- name: Add several ip addresses
```
`check_mode` (meaning: file isn't created/updated when check mode is enabled) and idempotence (meaning: changed is `False` when a task is repeated) should be checked.

pilou- commented on lib/ansible/modules/database/postgresql/pg_hba.py:
```diff
@@ -0,0 +1,598 @@
+  contype:
+    description:
+      - Type of the rule. Use emptystring C('') if you don't want to change file, but only want to read contents.
+    required: true
```
`contype` is required here but isn't required in `argument_spec`.

pilou- commented on lib/ansible/modules/database/postgresql/pg_hba.py:
```diff
@@ -0,0 +1,598 @@
+    module = AnsibleModule(
+        argument_spec=dict(
+            contype=dict(default=None, choices=PG_HBA_TYPES + ['']),
+            backup=dict(default=False, type='bool'),
```
`backup` is already a file common arguments ([see](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/basic.py#L240)).

sebasmannem commented:
@pilou- : I think I fixed all of your comments now. And shippable tests have succeeded too.
But it is a bit of a mess now (33 commits, some of the with rather personal remarks).
Should I merge all commits in a single commit? Like amend 383bfdfb515983136f38e2442c5ac639e3dac806?

pilou- commented on test/integration/targets/postgresql/tasks/pg_hba.yml:
```diff
@@ -0,0 +1,72 @@
+    order=sud
+    state=present
+  register: pg_hba_idempotency_check
+- assert:
```
The four `assert` could be grouped (`that` parameter is a list).

sebasmannem replied:
I had split this up, for better clarity which assert actually breaks.
But I merged the again.

pilou- commented on test/integration/targets/postgresql/tasks/pg_hba.yml:
```diff
@@ -0,0 +1,72 @@
+- assert:
+    that:
+    - 'pg_hba_checkmode_check.stat.exists == false'
+- debug:
```
`debug` tasks should be removed.

sebasmannem replied:
Fixed.

pilou- commented on test/integration/targets/postgresql/tasks/pg_hba.yml:
```diff
@@ -0,0 +1,72 @@
+- name: Make sure file does not exist
+  file:
```
YAML syntax should be used (`dest: /tmp/pg_hbp.conf`).

sebasmannem replied:
Fixed.

pilou- commented on test/integration/targets/postgresql/tasks/pg_hba.yml:
```diff
@@ -0,0 +1,72 @@
+    dest=/tmp/pg_hba.conf
+  register: pg_hba
+- name: Add ip again for idempotency check
+  pg_hba:
```
Idempotency test: the four `source`/`netmask` pairs could be tested.

sebasmannem replied:
Fixed.

pilou- commented on test/integration/targets/postgresql/tasks/pg_hba.yml:
```diff
@@ -0,0 +1,72 @@
+  stat:
+    path=/tmp/pg_hba.conf
+  register: pg_hba_checkmode_check
+- name: Add several ip addresses
```
Removing pg_hba entry must be tested too (idempotency, check mode).

sebasmannem replied:
Fixed.

pilou- commented on test/integration/targets/postgresql/tasks/pg_hba.yml:
```diff
@@ -0,0 +1,72 @@
+    source={{item.source}}
+    netmask={{item.netmask}}
+    method=md5
+    backup=true
```
backup file existence and content should be checked.

sebasmannem replied:
Fixed.

pilou- commented on lib/ansible/modules/database/postgresql/pg_hba.py:
```diff
@@ -0,0 +1,597 @@
+
+def rule2source_ipnw(rule):
+    '''
+    This method is used to get the source of a rul as an ipaddress object if possible.
```
typo s/rul/rule/

sebasmannem replied:
Fixed.

pilou- commented on lib/ansible/modules/database/postgresql/pg_hba.py:
```diff
@@ -0,0 +1,597 @@
+
+    def render(self):
+        '''
+        This method renders the content of the PgHba rules and commentsselfself.
```
typo: `commentsselfself`

sebasmannem replied:
Fixed.

pilou- commented on lib/ansible/modules/database/postgresql/pg_hba.py:
```diff
@@ -0,0 +1,597 @@
+            for k in ekeys:
+                if oldrule[k] != rule[k]:
+                    raise Exception('')
+        except Exception:
```
If the only expected exception type is the one raised line 390, a custom exception class should be used.

Note that it seems that a simpler way (without exception) could be used:
```
key = rule2key(rule)
oldrule = self.rules[key]
ekeys = set(list(oldrule.keys()) + list(rule.keys()))
ekeys.remove('line')
for k in ekeys:
    if oldrule[k] != rule[k]:
        self.rules[key] = rule
        self.diff['after']['pg_hba'].append(rule['line'])
        [...]
```

sebasmannem replied:
This is not the only exception. 
key = rule2key(rule)
Could also raise an error.
But I have made this part more descriptive.

pilou- commented on lib/ansible/modules/database/postgresql/pg_hba.py:
```diff
@@ -0,0 +1,597 @@
+
+    def is_empty(self):
+        '''
+        This methid can be used to check if a PgHba object is empty (no rules).
```
typo: `methid`

sebasmannem replied:
Fixed.

pilou- commented on lib/ansible/modules/database/postgresql/pg_hba.py:
```diff
@@ -0,0 +1,597 @@
+
+        return tuple(ret)
+
+    def is_empty(self):
```
This method seems unused, it must be removed.

sebasmannem replied:
Removed.

pilou- commented on lib/ansible/modules/database/postgresql/pg_hba.py:
```diff
@@ -0,0 +1,597 @@
+except ImportError:
+    HAS_IPADDRESS = False
+import copy
+from ansible.module_utils.basic import AnsibleModule, tempfile, shutil
```
`tempfile` and `shutil` are standard modules, use `import tempfile` and `import shutil` instead.

sebasmannem replied:
Fixed.

pilou- commented on lib/ansible/modules/database/postgresql/pg_hba.py:
```diff
@@ -0,0 +1,597 @@
+requirements:
+    - "ipaddress"
+
+author: Sebastiaan Mannem (github.com/sebasmannem)
```
Use `author: Sebastiaan Mannem (@sebasmannem)`

sebasmannem replied:
Fixed.

pilou- commented on lib/ansible/modules/database/postgresql/pg_hba.py:
```diff
@@ -0,0 +1,597 @@
+  order:
+    description:
+      - The entries will be written out in a specific order.
+      - With this option you can control by wich field they are ordered first, second and last.
```
typo: `wich`

sebasmannem replied:
Fixed.

pilou- commented on lib/ansible/modules/database/postgresql/pg_hba.py:
```diff
@@ -0,0 +1,597 @@
+      - Authentication method to be used.
+    default: "md5"
+    choices: [ "trust", "reject", "md5", "password", "gss", "sspi", "krb5", "ident", "peer", "ldap", "radius", "cert", "pam" ]
+  netmask:
```
With this example:
```
pg_hba:
  contype: host
  dest: /tmp/pg_hba.conf
  method: md5
  netmask: 172.21.0.0/24
  src: all
  users: pilou
  databases: awx
```
two entries are created: 
```
host    awx             pilou             172.21.0.0/24           md5
host    awx pilou   all md5
```
It seems an error should be raised when `netmask` is used with `src: all` or `src: samehost` or `src: samenet`.

pilou- replied:
Using the previous example, now an exception is raised:
```
TASK [pg_hba] ***
fatal: [localhost]: FAILED! => {
    "changed": false,
    "rc": 1
}

MSG:

MODULE FAILURE


MODULE_STDOUT:

Traceback (most recent call last):
  File "/tmp/ansible_FtGIVa/ansible_module_pg_hba.py", line 600, in <module>
    main()
  File "/tmp/ansible_FtGIVa/ansible_module_pg_hba.py", line 574, in main
    for rule in new_rules(contype, databases, users, source, netmask, method, options):
  File "/tmp/ansible_FtGIVa/ansible_module_pg_hba.py", line 518, in new_rules
    rule['src'] = str(rule2source_ipnw(rule))
  File "/tmp/ansible_FtGIVa/ansible_module_pg_hba.py", line 464, in rule2source_ipnw
    mask_as_ip = ipaddress.ip_address(u'{0}'.format(rule['mask']))
  File "/usr/lib/python2.7/dist-packages/ipaddress.py", line 166, in ip_address
    address)
ValueError: u'172.21.0.0/24' does not appear to be an IPv4 or IPv6 address



MODULE_STDERR:

Shared connection to localhost closed.
```
In order to check if `netmask` is used with `src: all` or `src: samehost` or `src: samenet`, should not a test be added in `main` method ?

pilou- commented on lib/ansible/modules/database/postgresql/pg_hba.py:
```diff
@@ -0,0 +1,597 @@
+      - s=source, d=databases, u=users.
+    default: "sdu"
+    choices: [ "sdu", "sud", "dsu", "dus", "usd", "uds" ]
+  src:
```
Should not this parameter named `address` ?

sebasmannem replied:
Changed it to address and added src and source as aliases.

pilou- commented on test/integration/targets/postgresql/tasks/pg_hba.yml:
```diff
@@ -0,0 +1,72 @@
+    var=pg_hba_idempotency_check
+- assert:
+    that:
+    - 'not pg_hba_idempotency_check is changed'
```
Check that
```
src: 172.21.0.0
netmask: 255.255.255.0
```
and
```
src: 172.21.0.0/24
```
generate the same configuration.

sebasmannem replied:
Added netmask_sameas_prefix

pilou- commented on lib/ansible/modules/database/postgresql/pg_hba.py:
```diff
@@ -0,0 +1,597 @@
+    PgHba object to read/write entries to/from.
+    pg_hba_file - the pg_hba file almost always /etc/pg_hba
+    """
+    def __init__(self, pg_hba_file=None, order="sdu", backup=False, create=False):
```
value of keyword parameter `order` is already `sdu` because `sdu` is the default value for `order` module parameter. Same for `backup` and `create`.

sebasmannem replied:
That is true. I made this so you can also import it as a module for something else. 
In that case I cannot be sure that the caller uses these defaults.
I'd like to keep it unless you're really bothered with it.

pilou- replied:
Ok, no need to change.

pilou- commented on lib/ansible/modules/database/postgresql/pg_hba.py:
```diff
@@ -0,0 +1,597 @@
+    pg_hba_file - the pg_hba file almost always /etc/pg_hba
+    """
+    def __init__(self, pg_hba_file=None, order="sdu", backup=False, create=False):
+        if order not in PG_HBA_ORDERS:
```
This test is always False since choices for `order` is already `PG_HBA_ORDERS`.

sebasmannem replied:
That is true. I made this so you can also import it as a module for something else. 
In that case I cannot be sure that the argument is an element of PG_HBA_ORDERS.
I'd like to keep it unless you're really bothered with it.

pilou- replied:
Ok, no need to change.

pilou- commented on lib/ansible/modules/database/postgresql/pg_hba.py:
```diff
@@ -0,0 +1,597 @@
+        So a single IP (/32) should have twice the weight of a /16 network.
+        To keep everything in the same weight scale,
+        - for ipv6, we use a weight scale of 0 (all possible ipv6 addresses) to 128 (single ip)
+        - for ipv4, we use a weight scale of 0 (all possible ipv6 addresses) to 128 (single ip)
```
typo: `ipv6`

sebasmannem replied:
Fixed.

pilou- commented on lib/ansible/modules/database/postgresql/pg_hba.py:
```diff
@@ -0,0 +1,597 @@
+        rule['line'] = line
+        return rule
+
+    def rule2weight(self, rule):
```
Using `order: sdu`, the file below was generated:
```
local   all             postgres                                peer
host    all             all             127.0.0.1/32            md5
local   all             all                                     peer
host    all             all             ::1/128                 md5
```
should not that be
```
local   all             postgres                                peer
local   all             all                                     peer
host    all             all             127.0.0.1/32            md5
host    all             all             ::1/128                 md5
``` 
?

sebasmannem replied:
Although this is technically not important, I fixed this.
I also added some better testcases to pg_hba_test_ips to better test this.

pilou- commented on test/integration/targets/postgresql/tasks/pg_hba.yml:
```diff
@@ -0,0 +1,133 @@
+    users: "{{item.users|default('all')}}"
+  with_items: "{{pg_hba_test_ips}}"
+  register: pg_hba_idempotency_check1
+- name: idemptotency not creating file check
```
typo: `idemptotency`

sebasmannem replied:
Fixed

pilou- commented on test/integration/targets/postgresql/tasks/pg_hba.yml:
```diff
@@ -0,0 +1,133 @@
+  stat:
+    path: /tmp/pg_hba.conf
+  register: pg_hba_checkmode_check
+- name: Remove several ip addresses for idempotency check
```
Because `/tmp/pg_hbd.conf` doesn't exist and `state` is `absent`, should not the module return `changed == False` (meaning: `pg_hba_idempotency_check1 is not changed`) ?

sebasmannem replied:
This was already the case, but because 'changed: true'. This now also is the case with changed: False.

sebasmannem commented:
> In order to check if netmask is used with src: all or src: samehost or src: samenet, should not a test be added in main method ?
Instead of a Traceback, the module now returns a clean error with module.fail.
To implement this, I have changed all raises from generic errors to subexceptions of PgHbaException.
Furthermore, in main(), I have used some try/excepts to catch PgHbaException exceptions and run module.fail, returning the message.

sebasmannem commented:
I have added a PgHbaRule object. I think it is a improvement.

sebasmannem commented:
@pilou- 
I cannot comment directly to your comment about
> Using the previous example, now an exception is raised:
but I fixed this and all other comments. Today I double checked textually and made some minor changes in remarks, but I think it is all done.

kustodian commented:
Would it make sense to rename this module to `postgres_hba`, so that it has the same prefix as all Postgres modules?

Andersson007 commented:
@sebasmannem Hello, I have a couple of questions for you:
1. Can you complete this module if we make full reviews including manual tests?
2. If you don't have enough time to do it soon, can anyone rewrite and complete this module later keeping your as an author?
(the first point is the best choice because you are an author)

sebasmannem commented:
Hi, I have updated my pg_hba module:

- Fixed a documentary thing (examples where broken / incomplete)
- Renamed the module to postgresql_pg_hba
- updated the version_added to 2.8
- Added scram to methods
Let's see what checks think, and after that I will bring it for review.

sebasmannem commented:
@Andersson007 : Do you happen to have any idea on why shippable is not running any checks?
Should somebody enable something?
Used to work up until May 2018.

Andersson007 commented:
@sebasmannem I don't really know. Sometimes it happens. I think it depends on servers load. It will be handled later.

Andersson007 commented:
you can check it locally before pushing, I'll sent you a short manual to your email during the next hour.

Andersson007 reviewed (changes requested):
**1. Formatting (these are recommendation from Ansible commiters):**

**1.0** Documentation = ''' => Documentation = r'''
Also for EXAMPLES and RETURN

Also we should add links to the official current documentation.

**1.1**
```
22 description:
23   - The fundamental function of the module is to create, or delete lines in pg_hba files.
24   - The lines in the file should be in a typical pg_hba form and lines should be unique per key (type, databases, users, source).
25     If they are not unique and the SID is 'the one to change', only one for C(state=present) or none for C(state=absent) of the SID's will remain.
26 extends_documentation_fragment: files
```
change it to (and so on to the end of the Documentation)
```
description:
- The fundamental function of the module is to create, or delete lines in pg_hba files.
- The lines in the file should be in a typical pg_hba form and lines should be unique per key (type, databases, users, source).
- If they are not unique and the SID is 'the one to change', only one for C(state=present) or none for C(state=absent) of the SID's will remain.
extends_documentation_fragment: files

```
```
90 notes:
91   - The default authentication assumes that on the host, you are either logging in as or
```
to
```
notes:
- The default authentication assumes that on the host, you are either logging in as or
```
also "requirements"

**1.2** I(type)=C(local) => I(type=local), also C(state=present) => I(state=present), etc.

**1.3**
```
35 aliases: [ 'source', 'src' ] => aliases: [ source, src ]
64 choices: [ "trust", "reject", "md5", "password", "gss", "sspi", "krb5", "ident", "peer", "ldap", "radius", "cert", "pam", "scram-sha-256" ] => ... needs to be without quotes + use the alphabetical order
51 choices: [ "local", "host", "hostssl", "hostnossl", "" ] => choices: [ host, hostssl, hostnossl, local ]
...
Seems it's None by default, so, you don't need to pass an empty string
```

**1.4** Add a type for each option and remove empty default values:
```
  options:
    description:
      - Additional options for the authentication I(method).
    default: ""
```
change to
```
69  options:
70    description:
71      - Additional options for the authentication I(method).
72    type: "str"
```

**1.5** IMO, needs the more detailed description
```
69  options:
70    description:
71      - Additional options for the authentication I(method).
72    default: ""
```

**1.6** Better to add more examples (with using each param; for detailed descriptions, use comments "# ..." above "- name:" where needed.

**1.7** If there are mutually exclusive params, needs to notice about it in the documentation and check in the code. There is a special method to do this.

**1.8** Needs to be changed
```
612    if not HAS_IPADDRESS:
613        module.fail_json(msg='Missing required libraries.')
```
to something like this:
```
    if not HAS_PSYCOPG2:
        module.fail_json(msg=missing_required_lib('psycopg2'), exception=PSYCOPG2_IMP_ERR)
```

**1.9** The common way is to use to_native(e):
```
633    try:
634        pg_hba = PgHba(dest, order, backup=backup, create=create)
635    except PgHbaError as error:
636        module.fail_json(msg='Error reading file:\n{0}'.format(error))
```
for example:
```
    except Exception as e:
        module.fail_json(msg="unable to connect to database: %s" % to_native(e))
```

===
**2. The code & logic**

**2.1** Maybe change the option backup: yes|no, to backup_file: '/path/to/file', type: 'path' ?

**2.2** Maybe better to add a warning if the file doesn't exist?
```
42  create:
43    description:
44      - Create an C(pg_hba) file if none exists.
45      - When set to false, an error is raised when the C(pg_hba) file doesn't exist.
46    default: false
47    type: bool
```

**2.3** Better check the database availability using psycopg2 to be able to reload the server configuration

**2.4** We discussed it, probably we should use psycopg2 to check the location of pg_hba.conf if dest is not passed.
```
56  dest:
57    description:
58      - Path to C(pg_hba) file to modify.
60    required: true
```

**2.5** Maybe to allow passing netmask with ipaddr like 10.10.10.1/32 ?
```
65  netmask:
66    description:
67      - The netmask of the source address.
68    default: ""
```

**2.6** Remove default=None/"", etc. from     module = AnsibleModule(argument_spec=dict(

**2.7** Devide tests by spaces, add ignore_errors and assertions after each test (IMO, this is easy to sort out), etc.
You can use for test/integration/targets/postgresql/tasks/postgresql_idx.yml or test/integration/targets/postgresql/tasks/postgresql_query.yml for anology

Also **your suggestions from our conversation sound good**, particularly:
- logs in using psycopg2.
- Detects PGDATA from that
- Changes pg_hba in that location (basically the way it currently works)
- Runs pg_reload_conf()
- Checks select * from pg_hba_file_rules where error is not null; (this is a PG10 feature, though)

Waiting for fixes and implementations.

Andersson007 commented on lib/ansible/modules/database/postgresql/postgresql_pg_hba.py:
```diff
@@ -0,0 +1,673 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+# Copyright: Ansible Project
```
```suggestion

# Copyright: (c) 2019, Sebastiaan Mannem (@sebasmannem) <YOURMAIL@HERE>
```

sebasmannem replied:
Resolved.

Andersson007 commented on lib/ansible/modules/database/postgresql/postgresql_pg_hba.py:
```diff
@@ -0,0 +1,673 @@
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = '''
```
```suggestion
DOCUMENTATION = r'''
```

Andersson007 commented on lib/ansible/modules/database/postgresql/postgresql_pg_hba.py:
```diff
@@ -0,0 +1,673 @@
+  options:
+    description:
+      - Additional options for the authentication I(method).
+    default: ""
```
```suggestion
    type: str
```
Seems it will empty by default

Andersson007 commented on lib/ansible/modules/database/postgresql/postgresql_pg_hba.py:
```diff
@@ -0,0 +1,673 @@
+    description:
+      - The lines will be added/modified when C(state=present) and removed when C(state=absent).
+    default: present
+    choices: [ "present", "absent" ]
```
```suggestion
    choices: [ absent, present ]
```
It's a strong recommendation to use alphabetic order for aliases, choices, etc

Andersson007 commented on lib/ansible/modules/database/postgresql/postgresql_pg_hba.py:
```diff
@@ -0,0 +1,673 @@
+      - With this option you can control by which field they are ordered first, second and last.
+      - s=source, d=databases, u=users.
+    default: "sdu"
+    choices: [ "sdu", "sud", "dsu", "dus", "usd", "uds" ]
```
```suggestion
    choices: [ dsu, dus, sdu, sud,  uds, usd]
```

Andersson007 replied:
I think you can use quotes if you want but several Ansible commiters always removed them from my PRs. Maybe it's their point of view but not really necessary.

Andersson007 commented on lib/ansible/modules/database/postgresql/postgresql_pg_hba.py:
```diff
@@ -0,0 +1,673 @@
+author: Sebastiaan Mannem (@sebasmannem)
+'''
+
+EXAMPLES = '''
```
```suggestion
EXAMPLES = r'''
```

Andersson007 commented on lib/ansible/modules/database/postgresql/postgresql_pg_hba.py:
```diff
@@ -0,0 +1,673 @@
+    source=192.168.0.100/24
+    databases=sales
+    method=md5
+
```
Better to add more examples, at least with removing rules too

Andersson007 commented on lib/ansible/modules/database/postgresql/postgresql_pg_hba.py:
```diff
@@ -0,0 +1,673 @@
+    method=trust
+'''
+
+RETURN = '''
```
```suggestion
RETURN = r'''
```

Andersson007 commented on test/integration/targets/postgresql/tasks/postgresql_pg_hba.yml:
```diff
@@ -0,0 +1,130 @@
+- name: Make sure file does not exist
+  file:
+    dest: /tmp/pg_hba.conf
+    state: absent
```
```suggestion
    state: absent
    
```
IMO, easier to read

Andersson007 commented on lib/ansible/modules/database/postgresql/postgresql_pg_hba.py:
```diff
@@ -0,0 +1,673 @@
+DOCUMENTATION = '''
+---
+module: postgresql_pg_hba
+short_description: Adds, removes or modifies an rule in a pg_hba file.
```
```suggestion
short_description: Add, remove or modify a rule in a pg_hba file
```
They usually use syntax of short description like this

Andersson007 commented on lib/ansible/modules/database/postgresql/postgresql_pg_hba.py:
```diff
@@ -0,0 +1,673 @@
+      - Will not be used for entries of I(type)=C(local).
+      - You can also use keywords C(all), C(samehost), and C(samenet).
+    default: "samehost"
+    aliases: [ 'source', 'src' ]
```
```suggestion
    type: str
    aliases: [ 'source', 'src' ]
```
It's a strong recommendation to use type for each option

Andersson007 commented on lib/ansible/modules/database/postgresql/postgresql_pg_hba.py:
```diff
@@ -0,0 +1,673 @@
+  contype:
+    description:
+      - Type of the rule. Use emptystring C('') if you don't want to change file, but only want to read contents.
+    choices: [ "local", "host", "hostssl", "hostnossl", "" ]
```
```suggestion
    choices: [ host, hostssl, hostnossl, local, "" ]
    type: str
```
type: str
alphabetic order
and is "" really neseccary here? seems if you don't pass it at all it will be "" by default

Andersson007 commented on lib/ansible/modules/database/postgresql/postgresql_pg_hba.py:
```diff
@@ -0,0 +1,673 @@
+    choices: [ "local", "host", "hostssl", "hostnossl", "" ]
+  databases:
+    description:
+      - Databases this line applies to.
```
```suggestion
      - Databases this line applies to.
      type: str
```

Andersson007 commented on lib/ansible/modules/database/postgresql/postgresql_pg_hba.py:
```diff
@@ -0,0 +1,673 @@
+    default: "all"
+  dest:
+    description:
+      - Path to C(pg_hba) file to modify.
```
```suggestion
      - Path to C(pg_hba) file to modify.
      type: path
```
or maybe str, it should be checked

Andersson007 commented on lib/ansible/modules/database/postgresql/postgresql_pg_hba.py:
```diff
@@ -0,0 +1,673 @@
+  method:
+    description:
+      - Authentication method to be used.
+    default: "md5"
```
```suggestion
    default: "md5"
    type: str
```

Andersson007 commented on lib/ansible/modules/database/postgresql/postgresql_pg_hba.py:
```diff
@@ -0,0 +1,673 @@
+    description:
+      - Authentication method to be used.
+    default: "md5"
+    choices: [ "trust", "reject", "md5", "password", "gss", "sspi", "krb5", "ident", "peer", "ldap", "radius", "cert", "pam", "scram-sha-256" ]
```
```suggestion
    choices: [ cert, gss, ident, krb5, ldap, md5, pam, password, peer, reject, radius, scram-sha-256, sspi, trust  ]
```
recommendation - alphabetic order

Andersson007 commented on lib/ansible/modules/database/postgresql/postgresql_pg_hba.py:
```diff
@@ -0,0 +1,673 @@
+    choices: [ "trust", "reject", "md5", "password", "gss", "sspi", "krb5", "ident", "peer", "ldap", "radius", "cert", "pam", "scram-sha-256" ]
+  netmask:
+    description:
+      - The netmask of the source address.
```
```suggestion
      - The netmask of the source address.
    type: str
```

Andersson007 commented on lib/ansible/modules/database/postgresql/postgresql_pg_hba.py:
```diff
@@ -0,0 +1,673 @@
+  netmask:
+    description:
+      - The netmask of the source address.
+    default: ""
```
```suggestion
```
and seems it will be empty by default, should check

Andersson007 commented on lib/ansible/modules/database/postgresql/postgresql_pg_hba.py:
```diff
@@ -0,0 +1,673 @@
+    choices: [ "present", "absent" ]
+  users:
+    description:
+      - Users this line applies to.
```
```suggestion
      - Users this line applies to.
    type: str
```

Andersson007 commented on lib/ansible/modules/database/postgresql/postgresql_pg_hba.py:
```diff
@@ -0,0 +1,673 @@
+   - The module supports a check mode and a diff mode.
+
+requirements:
+    - "ipaddress"
```
```suggestion
    - ipaddress
```

Andersson007 commented on lib/ansible/modules/database/postgresql/postgresql_pg_hba.py:
```diff
@@ -0,0 +1,673 @@
+    '''
+    This function is the main function of this module
+    '''
+    module = AnsibleModule(
```
We're changing this construction to using pg_utils from postgres lib like this [#52077](https://github.com/ansible/ansible/pull/52077/files):
```
    argument_spec = postgres_common_argument_spec()
    argument_spec.update(
        table=dict(type='str', required=True, aliases=['name']),
        state=dict(type='str', default="present", choices=["absent", "present"]),
        db=dict(type='str', default='', aliases=['login_db']),
        port=dict(type='int', default=5432, aliases=['login_port']),
        ssl_mode=dict(type='str', default='prefer', choices=['allow', 'disable', 'prefer', 'require', 'verify-ca', 'verify-full']),
        ssl_rootcert=dict(type='str'),
        tablespace=dict(type='str'),
        owner=dict(type='str'),
        unlogged=dict(type='bool'),
        like=dict(type='str'),
        including=dict(type='str'),
        rename=dict(type='str'),
        truncate=dict(type='bool'),
        columns=dict(type='list'),
        storage_params=dict(type='list'),
        session_role=dict(type='str'),
    )
    module = AnsibleModule(
        argument_spec=argument_spec,
        supports_check_mode=True,
    )
```

sebasmannem replied:
On this comment: I tried this, but it breaks, since postgresql_pg_hba does not have all of the postgres stuff to login, which all others do. We want to change that, but that is not in yet.
So I reverted this for now (commented the lines), and will add that when I implement the 'login and check' feature. Not sure if that will be in in this commit, bit that is fine. It can be added later.

Andersson007 replied:
Ok

Andersson007 commented on lib/ansible/modules/database/postgresql/postgresql_pg_hba.py:
```diff
@@ -0,0 +1,673 @@
+            databases=dict(type='str', default='all'),
+            dest=dict(type='path', required=True),
+            method=dict(type='str', default='md5', choices=PG_HBA_METHODS),
+            netmask=dict(type='str', default=''),
```
```suggestion
            netmask=dict(type='str'),
```
seems it's default='' by default

Andersson007 commented on lib/ansible/modules/database/postgresql/postgresql_pg_hba.py:
```diff
@@ -0,0 +1,673 @@
+            dest=dict(type='path', required=True),
+            method=dict(type='str', default='md5', choices=PG_HBA_METHODS),
+            netmask=dict(type='str', default=''),
+            options=dict(type='str', default=''),
```
```suggestion
            options=dict(type='str'),
```

Andersson007 commented on lib/ansible/modules/database/postgresql/postgresql_pg_hba.py:
```diff
@@ -0,0 +1,673 @@
+        supports_check_mode=True
+    )
+    if not HAS_IPADDRESS:
+        module.fail_json(msg='Missing required libraries.')
```
needs to be changed. the good example is a check for psycopg2
```
    if not HAS_PSYCOPG2:
        module.fail_json(msg=missing_required_lib('psycopg2'), exception=PSYCOPG2_IMP_ERR)
```

Andersson007 commented on lib/ansible/modules/database/postgresql/postgresql_pg_hba.py:
```diff
@@ -0,0 +1,673 @@
+      - You can also use keywords C(all), C(samehost), and C(samenet).
+    default: "samehost"
+    aliases: [ 'source', 'src' ]
+  backup:
```
Maybe change the option backup: yes|no, to backup_file: '/path/to/file', type: 'path' ?

sebasmannem replied:
I understand what you mean, but this is far more error prone. Do we have access to the path, permissions to write the file, etc. In addition to that, chances are high that they specify the same file for multiple runs, and keep overwriting their backup file.
Therefore I went for letting the module make files where it always can (/tmp) and generating new filenames for every backup.

sebasmannem replied:
Maybe add the option to set backup_file to a path if they want to hard specify it?

Andersson007 replied:
yup, the more flexible the better ;) and it's safer (e.g., unexpected reboot).
I think anyawy there will be an issue from people soon with a request to implement a possibility of choosing the bacup file path.

sebasmannem replied:
Yeah, I agree. Will try to add it. When should we commit this one?

Andersson007 replied:
Better before 3 March (4 feature freeze).
However if it's possible to add it later without changing the ui, that's OK.

sebasmannem replied:
It is in there.

Andersson007 commented on lib/ansible/modules/database/postgresql/postgresql_pg_hba.py:
```diff
@@ -0,0 +1,673 @@
+        The location of the backup is returned in the (backup) variable by this module.
+    default: false
+    type: bool
+  create:
```
IMO, better to add a warn if the file doesn't exist, just to inform the admin

sebasmannem replied:
That is the behaviour when create is set to false (default)...

Andersson007 replied:
Ok

Andersson007 commented:
Also your ideas were:
```
- logs in using psycopg2.
- Detects PGDATA from that
- Changes pg_hba in that location (basically the way it currently works)
- Runs pg_reload_conf()
- Checks select * from pg_hba_file_rules where error is not null; (this is a PG10 feature, though)
```
If you don't have enough time to implement these, we can leave the PR without them but they sound good.

IMO, you should implement now all that you can't implement in the future because we might not change all related to UI after it will released.

Waiting for your feedback for the review above.

Andersson007 commented:
@kostiantyn-nemchenko , hello, could you please look at this one ?

Commit:
Adding a backup_file option

sebasmannem commented:
> Also your ideas were:
> 
> ```
> - logs in using psycopg2.
> - Detects PGDATA from that
> - Changes pg_hba in that location (basically the way it currently works)
> - Runs pg_reload_conf()
> - Checks select * from pg_hba_file_rules where error is not null; (this is a PG10 feature, though)
> ```
> If you don't have enough time to implement these, we can leave the PR without them but they sound good.
> 
> IMO, you should implement now all that you can't implement in the future because we might not change all related to UI after it will released.
> 
> Waiting for your feedback for the review above.

Yeah, I am not sure if I can have this in rather fast enough.
this could be added in the next release though. No breaking changes there.
I will work on it, and let you know.

Andersson007 reviewed (approved):
shipit
